File: aud-twitter.py
#!/usr/bin/python
#
# aud-twitter: Publish on twitter what you are listening on Audacious
# based on mpd-twitter by Zerial
#

import os, subprocess, ConfigParser, sys
from twitter import Api

config = ConfigParser.ConfigParser()
params = config.read(['/etc/aud-twitter.conf', os.path.expanduser('~/.aud-twitter')])

CONSUMER_KEY = config.get('connect_params', 'CONSUMER_KEY', 1)
CONSUMER_SECRET = config.get('connect_params', 'CONSUMER_SECRET', 1)
ACCESS_TOKEN_KEY = config.get('connect_params', 'ACCESS_TOKEN_KEY', 1)
ACCESS_TOKEN_SECRET = config.get('connect_params', 'ACCESS_TOKEN_SECRET', 1)

TWEET_BEFORE = config.get('tweet_params', 'BEFORE', 1)
TWEET_AFTER = config.get('tweet_params', 'AFTER', 1)
TWEET_EXTRA = sys.argv[1:]

curr_song = subprocess.Popen(['audtool', 'current-song'], stdout=subprocess.PIPE)
status = TWEET_BEFORE + ' ' + curr_song.stdout.read().rstrip() + ' ' + TWEET_AFTER
if len(TWEET_EXTRA):
    status += '// ' + ' '.join(TWEET_EXTRA)

api = Api(consumer_key=CONSUMER_KEY, consumer_secret=CONSUMER_SECRET, access_token_key=ACCESS_TOKEN_KEY, access_token_secret=ACCESS_TOKEN_SECRET)
new_status = api.PostUpdate(status)

File: aud-twitter.pl
#!/usr/bin/perl
use strict;
use Config::Tiny;
use Net::Twitter;

my $config = Config::Tiny->read("$ENV{HOME}/.aud-twitter")
    or die "error reading config: @{[Config::Tiny->errstr]}\n";

my %oauth = map { $_ => ($config->{connect_params}->{$_} or die "invalid config: missing $_\n") }
    qw(consumer_key consumer_secret access_token access_token_secret);

my $tweet_before = $config->{tweet_params}->{before};
my $tweet_after = $config->{tweet_params}->{after};
my $tweet_extra = join ' ', @ARGV;

my $curr_song = `audtool current-song`;
chomp $curr_song;
die 'no current song' if !length($curr_song);

my $status = "$tweet_before $curr_song $tweet_after";
$status .= "// $tweet_extra" if length($tweet_extra);

print "sending tweet: $status\n";
my $api = Net::Twitter->new(traits => [qw/OAuth API::REST/], %oauth);
$api->update($status);
